src/wallet/crypter.cpp

```diff
@@ -205,6 +205,74 @@ bool CCryptoKeyStore::Unlock(const CKeyingMaterial& vMasterKeyIn)
     return true;
 }
 
+bool CCryptoKeyStore::AddCryptedPaperKey(const CKeyingMaterial& vchCryptedPaperKey)
+{
+    {
+        LOCK(cs_KeyStore);
+        if (!SetCrypted())
+            return false;
+
+        this->vchCryptedPaperKey = vchCryptedPaperKey;
+    }
+    return true;
+}
+
+bool CCryptoKeyStore::AddPaperKey(const SecureString& paperKey)
+{
+    {
+        LOCK(cs_KeyStore);
+        if (!IsCrypted())
+            return CBasicKeyStore::AddPaperKey(paperKey);
+
+        if (IsLocked())
+            return false;
+
+        std::vector<unsigned char> vchOut;
+        CKeyingMaterial vchPaperKey(paperKey.begin(), paperKey.end());
+        bool result = EncryptSecret(vMasterKey, vchPaperKey, DoubleHashOfString("paperkey"), vchOut);
```

Comment:
Is it really should be a constant as iv parameter? If yes then let's name it and remove DoubleHashOfString function from this class.

Reply:
Hi @sergey-yezovskikh ,

Yes, I agree with you. I've made it a static non-member function that I call only once (since the result is a constant for _"paperkey"_).

Additionally, I've refactored the function itself.

Finally, the paper key code was merged from the **paicoin-feature-investor** branch and maybe we can add the changes in this commit there as well.

Commit: https://github.com/projectpai/paicoin/pull/102/commits/3a4fd5528646885aabb12dd946510c5b58cc890d